noir-projects/noir-contracts/contracts/amm_contract/src/main.nr

```diff
@@ -0,0 +1,531 @@
+
+        // With the current balances known, we can calculate the token amounts to the pool, respecting the user's
+        // minimum deposit preferences.
+        let (amount0, amount1) = get_amounts_to_add(
+            amount0_max,
+            amount1_max,
+            amount0_min,
+            amount1_min,
+            balance0,
+            balance1,
+        );
+
+        // Return any excess from the original token deposits.
+        let refund_amount_token0 = amount0_max - amount0;
+        let refund_amount_token1 = amount1_max - amount1;
+
+        // We can simply skip the refund if the amount to return is 0 in order to save gas: the partial note will
+        // simply stay in public storage and not be completed, but this is not an issue.
+        if (refund_amount_token0 > U128::zero()) {
+            token0
+                .finalize_transfer_to_private(
+                    refund_amount_token0.to_integer(),
+                    refund_token0_hiding_point_slot,
+                )
+                .call(&mut context);
+        }
+        if (refund_amount_token1 > U128::zero()) {
+            token1
+                .finalize_transfer_to_private(
+                    refund_amount_token1.to_integer(),
+                    refund_token1_hiding_point_slot,
+                )
+                .call(&mut context);
+        }
+
+        // With the deposit amounts known, we can compute the number of liquidity tokens to mint and finalize the
+        // depositor's partial note.
+        let total_supply = U128::from_integer(liquidity_token.total_supply().view(&mut context));
+        let liquidity_amount = if total_supply != U128::zero() {
+            // The liquidity token supply increases by the same ratio as the balances. In case one of the token balances
+            // increased with a ratio different from the other one, we simply take the smallest value.
+            std::cmp::min(
+                (amount0 * total_supply) / balance0,
+                (amount1 * total_supply) / balance1,
+            )
+        } else {
+            // The zero total supply case (i.e. pool initialization) is special as we can't increase the supply
+            // proportionally. We instead set the initial liquidity to an arbitrary amount.
+            // We could set the initial liquidity to be equal to the pool invariant (i.e. sqrt(amount0 * amount1)) if
+            // we wanted to collect protocol fees over swap fees (in the style of Uniswap v2), but we choose not to in
+            // order to keep things simple.
+
+            // As part of initialization, we mint some tokens to the zero address to 'lock' them (i.e. make them
+            // impossible to redeem), guaranteeing total supply will never be zero again.
+            liquidity_token
+                .mint_to_public(AztecAddress::zero(), MINIMUM_LIQUIDITY.to_integer())
+                .call(&mut context);
+
+            INITIAL_LIQUIDITY
```

Comment:
should this be `sqrt(amount0 * amount1)`? You can grab sqrt function from here https://github.com/olehmisar/nodash/blob/24b645692ac3d3fb2e3a5ac0f43314e362294a84/src/math/sqrt.nr

Reply:
There's an explanation right above this covering why we set an arbitrary amount.

Reply:
Also, that version of `sqrt` is meant to be called in a circuit, not in the AVM. And I'd not want to introduce an external dependency just for this.